drivers/src/gicv2.rs

```diff
@@ -0,0 +1,247 @@
+use tock_registers::interfaces::{ReadWriteable, Readable, Writeable};
+use tock_registers::register_bitfields;
+use tock_registers::registers::{ReadOnly, ReadWrite};
+
+pub struct GicV2 {
+    pub distributor: &'static GicDistributor,
+    pub cpu: &'static GicCpu,
+}
+
+impl GicV2 {
+    pub fn new(distributor_base: usize, cpu_base: usize) -> Self {
+        let distributor = unsafe {
+            (distributor_base as *const GicDistributor)
+                .as_ref()
+                .unwrap()
+        };
+        let cpu = unsafe { (cpu_base as *const GicCpu).as_ref().unwrap() };
+        let mut gic = Self { distributor, cpu };
+
+        gic.init_distributor();
+
+        gic
+    }
+
+    pub fn disable_interrupts(&mut self) {
+        self.distributor
+            .CTLR
+            .modify(GICD_CTLR::EnableGrp0::Disable + GICD_CTLR::EnableGrp1::Disable);
+    }
+
+    pub fn enable_interrupts(&mut self) {
+        self.distributor
+            .CTLR
+            .modify(GICD_CTLR::EnableGrp0::Enable + GICD_CTLR::EnableGrp1::Enable);
+    }
+
+    pub fn nlines(&self) -> usize {
+        let n = self.distributor.TYPER.read(GICD_TYPER::ITLinesNumber) as usize;
+
+        32 * (n + 1)
+    }
+
+    /// Put the Gic in a known state.
+    fn init_distributor(&mut self) {
+        self.disable_interrupts();
+
+        for i in 0..(self.nlines() / 32) {
+            // Disable all interrupts.
+            // Each bit corresponds to a line, writing 1 disables forwarding of the corresponding
+            // interrupt line.
+            self.distributor.ICENABLER[i].set(0xffff_ffff);
+
+            // Clear pending interrupts.
+            self.distributor.ICPENDR[i].set(0xffff_ffff);
+
+            // Clear active interrupts.
+            self.distributor.ICACTIVER[i].set(0xffff_ffff);
+        }
+
+        for i in 0..(self.nlines() / 4) {
+            // Targets all interrupts to core 0.
+            self.distributor.ITARGETSR[i].set(0x0101_0101);
+        }
+
+        for i in 1..(self.nlines() / 16) {
+            // Set all interrupts to level-triggered.
+            self.distributor.ICFGR[i].set(0);
+        }
+
+        // TODO: this should be moved somewhere else so other cores can run it.
```

Comment:
We can leave this for now but since I think a `static AtomicBool` could do the trick quite nicely

Reply:
Why ?